keystore/src/transaction.rs

```diff
@@ -80,379 +75,396 @@ impl EntityId {
+        transaction.commit_tx().await?;
+        Ok(entity)
     }
 
-    async fn find_all<E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection>>(
+    pub(crate) async fn remove<
+        E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection> + EntityTransactionExt,
+        S: AsRef<[u8]>,
+    >(
         &self,
-        params: EntityFindParams,
-    ) -> CryptoKeystoreResult<Vec<E>> {
-        self.conn.find_all(params).await
+        id: S,
+    ) -> CryptoKeystoreResult<()> {
+        let mut conn = self.cache.borrow_conn().await?;
+        #[cfg(target_family = "wasm")]
+        let transaction = conn.new_transaction(&[E::COLLECTION_NAME]).await?;
+        #[cfg(not(target_family = "wasm"))]
+        let transaction = conn.new_transaction().await?;
+        E::delete(&transaction, id.as_ref().into()).await?;
+        transaction.commit_tx().await?;
+        let mut deleted_list = self.deleted.write().await;
+        deleted_list.push(EntityId::from_collection_name(E::COLLECTION_NAME, id.as_ref())?);
+        Ok(())
     }
 
-    async fn find_many<E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection>>(
+    pub(crate) async fn child_groups<
+        E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection>
+            + crate::entities::PersistedMlsGroupExt
+            + Sync,
+    >(
         &self,
-        ids: &[Vec<u8>],
+        entity: E,
+        persisted_records: Vec<E>,
     ) -> CryptoKeystoreResult<Vec<E>> {
-        self.conn.find_many(ids).await
+        let mut conn = self.cache.borrow_conn().await?;
+        let cached_records = entity.child_groups(conn.deref_mut()).await?;
+        Ok(self
+            .merge_records(cached_records, persisted_records, EntityFindParams::default())
+            .await)
     }
 
-    async fn count<E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection>>(
-        &self,
-    ) -> CryptoKeystoreResult<usize> {
-        let mut conn = self.conn.borrow_conn().await?;
-        E::count(&mut conn).await
+    pub(crate) async fn cred_delete_by_credential(&self, cred: Vec<u8>) -> CryptoKeystoreResult<()> {
+        let mut conn = self.cache.borrow_conn().await?;
+        #[cfg(target_family = "wasm")]
+        let transaction = conn.new_transaction(&[MlsCredential::COLLECTION_NAME]).await?;
+        #[cfg(not(target_family = "wasm"))]
+        let transaction = conn.new_transaction().await?;
+        MlsCredential::delete_by_credential(&transaction, cred.clone()).await?;
+        transaction.commit_tx().await?;
+        let mut deleted_list = self.deleted_credentials.write().await;
+        deleted_list.push(cred);
+        Ok(())
     }
-}
 
-impl KeystoreTransaction {
-    pub fn new(conn: Connection) -> Self {
-        Self {
-            conn,
-            operations: Default::default(),
+    pub(crate) async fn find<E>(&self, id: &[u8]) -> CryptoKeystoreResult<Option<Option<E>>>
```

Comment:
Why do we have two `Option`s in a row in the result?

Reply:
Thanks for pointing out that this may need clarification! I'm adding the following docstring
```rust
/// The result of this function will have different contents for different scenarios: 
/// * `Some(Some(E))` - the transaction cache contains the record
/// * `Some(None)` - the deletion of the record has been cached
/// * `None` - there is no information about the record in the cache 
```